### ISSUE TITLE
Event type filtering causes consumers to miss records

### ISSUE THREAD

**Issue by andrashatvani** [spring-cloud/spring-cloud-stream#2627]
**Describe the issue**
With event type filtering enabled via the `KafkaStreamsExtendedBindingProperties` in application.yml several records matching the filters have been missed in a stream with 8 input topics consumed as either `KStreams` or `KTables`. According to my observation the issue wasn't present when the topic was consumed as a `GlobalKTable`.
Once I removed the `event-types` filters from the input topics all records have arrived at the stream.

**To Reproduce**
I couldn't reproduce it with an integration test, only in system test environments with thousands of records. Always different records with varying count from mostly, but not not only from the same topic have been missed. Approximately 20 out of 200k records have not arrived at the stream

**Version of the framework**
Spring Cloud Stream 2022.0.0 with Sprint Boot 3.0.1

**Expected behavior**
All records matching the filter arrive at the stream.

**Screenshots**

**Additional context**
It might be a race condition or other concurrency issue in the event type filtering engine.
The event type filtering has entirely been deactivated in all services to avoid data loss.

**Comment by sobychacko** [comment]
@andrashatvani Thanks for reporting the issue. It is hard to reproduce this issue on our end since it only happens under heavy loads. Do you set concurrency in the application? What is your setting? Please let us know if you can devise some deterministic ways to reproduce the issue.

**Comment by andrashatvani** [comment]
Yes, I've set `spring.cloud.stream.kafka.streams.binder.configuration.num.stream.threads: 24` (one stream thread per CPU core).
Furthermore, I can provide the stream configuration where it happened on each run:

```kotlin
@Configuration
class ProjektAggregatorStreamConfiguration {
    @Bean
    fun projektAggregator() =
        Function { projekte: KStream<String, ProjektEvent> ->
            Function { projektstatus: KStream<String, ProjektStatusEvent> ->
                Function { befunde: KStream<String, ProjektBefundAggregat> ->
                    Function { aufgaben: KStream<String, ProjektAufgabeAggregat> ->
                        Function { wirtschaftseinheiten: GlobalKTable<String, WirtschaftseinheitAggregat> ->
                            Function { durchfuehrungen: KStream<String, ProjektDurchfuehrungAggregat> ->
                                Function { gruppen: KStream<String, ProjektGruppeAggregat> ->
                                    Function { mietobjekte: GlobalKTable<String, MietobjektAggregat> ->
                                        projekte
                                            .filterNot { _, projektEvent -> projektEvent.action == CREATE_REQUEST }
                                            .leftJoin(wirtschaftseinheiten)
                                            .leftJoin(mietobjekte)
                                            .cogroup { _, current, previous: ProjektAggregat ->
                                                previous.copy(
                                                    projekt = current.projekt,
                                                    wirtschaftseinheit = current.wirtschaftseinheit,
                                                    mietobjekt = current.mietobjekt,
                                                    projektErstelltAm = current.projektErstelltAm
                                                )
                                            }
                                            .cogroup(projektstatus.groupByKey()) { _, projektstatusEvent, aggregat -> aggregat + projektstatusEvent }
                                            .cogroup(befunde.groupByKey()) { _, befundAggregat, aggregat -> aggregat + befundAggregat }
                                            .cogroup(aufgaben.groupByKey()) { _, aufgabeAggregat, aggregat -> aggregat + aufgabeAggregat }
                                            .cogroup(durchfuehrungen.groupByKey()) { _, durchfuehrungAggregat, aggregat -> aggregat + durchfuehrungAggregat }
                                            .cogroup(gruppen.groupByKey()) { _, gruppeAggregat, aggregat -> aggregat + gruppeAggregat }
                                            .aggregate({ ProjektAggregat() }, Materialized.`as`(ProjektStore.NAME))
                                            .toStream()
                                            .filterNot { _, projektAggregat -> projektAggregat.projekt == null }
                                            .process(ProcessorSupplier { EventTypeHeaderProcessor() })
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
}
```
The records were missed mostly in, but not limited to the `aufgaben` and `durchfuehrungen` streams which contain ca. 100k records each in compacted topics.
Changing cogroups to joins didn't have any positive effect.

**Comment by sobychacko** [comment]
@andrashatvani Is it consistently happening? Since you have some deterministic path to reproduce the issue, I wonder if you could debug in the binder code and come up with some insights?

**Comment by sobychacko** [comment]
Closing the issue. Feel free to re-open if you face further issues.

**Comment by sobychacko** [comment]
@andrashatvani We probably have fixed this issue via [another issue](https://github.com/spring-cloud/spring-cloud-stream/issues/2823). That fix is in `4.1.x` and `4.0.x` (`4.0.5-SNAPSHOT`) versions.
